File: client/libs/raycast.ts
import { Coords3 } from './types';

type GetVoxel = (vx: number, vy: number, vz: number, wx?: number, wy?: number, wz?: number) => boolean;

function traceRay(
  getVoxel: GetVoxel,
  px: number,
  py: number,
  pz: number,
  dx: number,
  dy: number,
  dz: number,
  maxD: number,
  hitPos: Coords3,
  hitNorm: Coords3,
) {
  // consider raycast vector to be parametrized by t
  //   vec = [px,py,pz] + t * [dx,dy,dz]

  // algo below is as described by this paper:
  // http://www.cse.chalmers.se/edu/year/2010/course/TDA361/grid.pdf

  let t = 0.0,
    ix = Math.floor(px) | 0,
    iy = Math.floor(py) | 0,
    iz = Math.floor(pz) | 0;
  const stepX = dx > 0 ? 1 : -1,
    stepY = dy > 0 ? 1 : -1,
    stepZ = dz > 0 ? 1 : -1;
  // dx,dy,dz are already normalized
  const txDelta = Math.abs(1 / dx),
    tyDelta = Math.abs(1 / dy),
    tzDelta = Math.abs(1 / dz),
    xDist = stepX > 0 ? ix + 1 - px : px - ix,
    yDist = stepY > 0 ? iy + 1 - py : py - iy,
    zDist = stepZ > 0 ? iz + 1 - pz : pz - iz;
  // location of nearest voxel boundary, in units of t
  let txMax = txDelta < Infinity ? txDelta * xDist : Infinity,
    tyMax = tyDelta < Infinity ? tyDelta * yDist : Infinity,
    tzMax = tzDelta < Infinity ? tzDelta * zDist : Infinity,
    steppedIndex = -1;

  // main loop along raycast vector
  while (t <= maxD) {
    const hx = px + t * dx;
    const hy = py + t * dy;
    const hz = pz + t * dz;

    // exit check
    const b = getVoxel(ix, iy, iz, hx, hy, hz);
    if (b) {
      if (hitPos) {
        hitPos[0] = hx;
        hitPos[1] = hy;
        hitPos[2] = hz;
      }
      if (hitNorm) {
        hitNorm[0] = hitNorm[1] = hitNorm[2] = 0;
        if (steppedIndex === 0) hitNorm[0] = -stepX;
        if (steppedIndex === 1) hitNorm[1] = -stepY;
        if (steppedIndex === 2) hitNorm[2] = -stepZ;
      }
      return b;
    }

    // advance t to next nearest voxel boundary
    if (txMax < tyMax) {
      if (txMax < tzMax) {
        ix += stepX;
        t = txMax;
        txMax += txDelta;
        steppedIndex = 0;
      } else {
        iz += stepZ;
        t = tzMax;
        tzMax += tzDelta;
        steppedIndex = 2;
      }
    } else {
      if (tyMax < tzMax) {
        iy += stepY;
        t = tyMax;
        tyMax += tyDelta;
        steppedIndex = 1;
      } else {
        iz += stepZ;
        t = tzMax;
        tzMax += tzDelta;
        steppedIndex = 2;
      }
    }
  }

  // no voxel hit found
  if (hitPos) {
    hitPos[0] = px + t * dx;
    hitPos[1] = py + t * dy;
    hitPos[2] = pz + t * dz;
  }
  if (hitNorm) {
    hitNorm[0] = hitNorm[1] = hitNorm[2] = 0;
  }

  return 0;
}

// conform inputs

function raycast(
  getVoxel: GetVoxel,
  origin: Coords3,
  direction: Coords3,
  maxD: number,
  hitPos: Coords3,
  hitNorm: Coords3,
) {
  const px = +origin[0];
  const py = +origin[1];
  const pz = +origin[2];

  let dx = +direction[0];
  let dy = +direction[1];
  let dz = +direction[2];
  const ds = Math.sqrt(dx * dx + dy * dy + dz * dz);

  if (ds === 0) {
    throw new Error("Can't raycast along a zero vector");
  }

  dx /= ds;
  dy /= ds;
  dz /= ds;

  return traceRay(getVoxel, px, py, pz, dx, dy, dz, maxD, hitPos, hitNorm);
}

export { raycast };

File: server/utils/src/raycast.rs
#![allow(dead_code)]

// HELP FROM https://github.com/andyhall/fast-voxel-raycast/blob/master/index.js

use server_common::{math::approx_equals, types::GetVoxel, vec::Vec3};

#[allow(clippy::too_many_arguments)]
fn trace_ray(
    get_voxel: GetVoxel,
    px: f32,
    py: f32,
    pz: f32,
    dx: f32,
    dy: f32,
    dz: f32,
    max_d: f32,
    hit_pos: &mut Vec3<f32>,
    hit_norm: &mut Vec3<i32>,
) -> bool {
    let mut t = 0.0;

    let mut ix = px.floor() as i32;
    let mut iy = py.floor() as i32;
    let mut iz = pz.floor() as i32;

    let step_x = if dx > 0.0 { 1 } else { -1 };
    let step_y = if dy > 0.0 { 1 } else { -1 };
    let step_z = if dz > 0.0 { 1 } else { -1 };

    let tx_delta = (1.0 / dx).abs();
    let ty_delta = (1.0 / dy).abs();
    let tz_delta = (1.0 / dz).abs();

    let x_dist = if step_x > 0 {
        ix as f32 + 1.0 - px
    } else {
        px - ix as f32
    };
    let y_dist = if step_y > 0 {
        iy as f32 + 1.0 - py
    } else {
        py - iy as f32
    };
    let z_dist = if step_z > 0 {
        iz as f32 + 1.0 - pz
    } else {
        pz - iz as f32
    };

    let mut tx_max = if tx_delta < f32::MAX {
        tx_delta * x_dist
    } else {
        f32::MAX
    };
    let mut ty_max = if ty_delta < f32::MAX {
        ty_delta * y_dist
    } else {
        f32::MAX
    };
    let mut tz_max = if tz_delta < f32::MAX {
        tz_delta * z_dist
    } else {
        f32::MAX
    };

    let mut stepped_index = -1;

    #[allow(clippy::while_immutable_condition, clippy::collapsible_else_if)]
    while t <= max_d {
        // exit check
        let v = get_voxel(ix, iy, iz);
        if v {
            hit_pos.0 = px + t as f32 * dx;
            hit_pos.1 = py + t as f32 * dy;
            hit_pos.2 = pz + t as f32 * dz;

            hit_norm.0 = 0;
            hit_norm.1 = 0;
            hit_norm.2 = 0;

            if stepped_index == 0 {
                hit_norm.0 = -step_x;
            } else if stepped_index == 1 {
                hit_norm.1 = -step_y;
            } else if stepped_index == 2 {
                hit_norm.2 = -step_z;
            }

            return v;
        }

        if tx_max < ty_max {
            if tx_max < tz_max {
                ix += step_x;
                t = tx_max;
                tx_max += tx_delta;
                stepped_index = 0;
            } else {
                iz += step_z;
                t = tz_max;
                tz_max += tz_delta;
                stepped_index = 2;
            }
        } else {
            if ty_max < tz_max {
                iy += step_y;
                t = ty_max;
                ty_max += ty_delta;
                stepped_index = 1;
            } else {
                iz += step_z;
                t = tz_max;
                tz_max += tz_delta;
                stepped_index = 2;
            }
        }
    }

    // no voxel hit found
    hit_pos.0 = px + t * dx;
    hit_pos.1 = py + t * dy;
    hit_pos.2 = pz + t * dz;

    hit_norm.0 = 0;
    hit_norm.1 = 0;
    hit_norm.2 = 0;

    false
}

pub fn trace(
    max_d: f32,
    get_voxel: GetVoxel,
    origin: &mut Vec3<f32>,
    direction: &mut Vec3<f32>,
    hit_pos: &mut Vec3<f32>,
    hit_norm: &mut Vec3<i32>,
) -> bool {
    let Vec3(px, py, pz) = origin;
    let Vec3(dx, dy, dz) = direction;
    let ds = (*dx * *dx + *dy * *dy + *dz * *dz).sqrt();

    if approx_equals(&ds, &0.0) {
        // ?should return an error?
        panic!("Can't raycast along a zero vector");
    }

    *dx /= ds;
    *dy /= ds;
    *dz /= ds;

    trace_ray(
        get_voxel, *px, *py, *pz, *dx, *dy, *dz, max_d, hit_pos, hit_norm,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sphere_test() {
        let voxel = |x: i32, y: i32, z: i32| x == 0 && y == 0 && z == 0;

        let mut hit_position = Vec3::default();
        let mut hit_normal = Vec3::default();

        // 2 * PI
        for t100 in (0..628).step_by(25) {
            let theta = t100 as f32 / 100.0;
            for p100 in (0..628).step_by(25) {
                let phi = p100 as f32 / 100.0;
                let mut dir = Vec3(theta.cos() * phi.cos(), theta.sin() * phi.cos(), phi.sin());
                for x1000 in (1..1000).step_by(250) {
                    let x = x1000 as f32 / 1000.0;
                    for y1000 in (1..1000).step_by(250) {
                        let y = y1000 as f32 / 1000.0;
                        for z1000 in (1..1000).step_by(250) {
                            let z = z1000 as f32 / 1000.0;
                            let mut pos = Vec3(x - 2.0 * dir.0, y - 2.0 * dir.1, z - 2.0 * dir.2);
                            let b = trace(
                                10.0,
                                &voxel,
                                &mut pos,
                                &mut dir,
                                &mut hit_position,
                                &mut hit_normal,
                            );
                            assert!(b);
                        }
                    }
                }
            }
        }
    }
}
